File: python/tests/test_kinodynamic_rrt_star.py
"""Kinodynamic RRT* (Webb & van den Berg 2013): finds a dynamically-feasible,
collision-free path to the goal, fails gracefully when walled off, validates its
params, and its optimal-steering cost obeys the metric contract."""

from __future__ import annotations

from pathlib import Path

import pytest
import yaml
from conftest import CONFIG_DIR, grid_from, open_grid

from navigation.core.params import ParamError, ParamSet
from navigation.global_planning.sampling.kinodynamic_rrt_star import (
    KinodynamicRRTStar,
    optimal_cost,
)
from navigation.maps.occupancy_grid import OccupancyGrid2D

_ALGO = "kinodynamic_rrt_star"


def _capped_config(tmp_path: Path, **overrides: float) -> ParamSet:
    """Real config with per-name overrides so tests stay fast and deterministic."""
    doc = yaml.safe_load((CONFIG_DIR / f"{_ALGO}.yaml").read_text())
    for entry in doc["params"]:
        if entry["name"] in overrides:
            entry["default"] = overrides[entry["name"]]
    out = tmp_path / f"{_ALGO}.yaml"
    out.write_text(yaml.safe_dump(doc), encoding="utf-8")
    return ParamSet.from_yaml(out)


def _motion_valid_path(grid: OccupancyGrid2D, path: list[tuple[float, float]]) -> bool:
    return all(grid.is_motion_valid(a, b) for a, b in zip(path, path[1:], strict=False))


def test_reaches_goal_with_feasible_collision_free_path(tmp_path: Path) -> None:
    grid = open_grid(10, 10, seed=7)
    start, goal = (0.75, 0.75), (9.0, 9.0)
    params = _capped_config(
        tmp_path, max_iterations=1500, goal_bias=0.2, goal_tolerance=1.5, seed=3
    )
    result = KinodynamicRRTStar(params).plan(grid, start, goal)
    assert result.success
    assert result.path[0] == pytest.approx(start)
    # Optimal steering arrives exactly at the goal rest-state; the projected endpoint
    # is within tolerance of the goal position.
    assert grid.distance(result.path[-1], goal) <= params.get_float("goal_tolerance")
    # The propagated double-integrator trajectory must be collision-free throughout.
    assert _motion_valid_path(grid, result.path)
    assert result.cost > 0.0


def test_no_path_when_walled_off(tmp_path: Path) -> None:
    # Fully occupied middle column: no trajectory can cross it.
    grid = grid_from(["..#..", "..#..", "..#..", "..#..", "..#.."], seed=3)
    params = _capped_config(tmp_path, max_iterations=300, goal_tolerance=1.0)
    result = KinodynamicRRTStar(params).plan(grid, (0.5, 0.5), (4.5, 4.5))
    assert not result.success
    assert result.path == []
    assert result.cost == 0.0


def test_out_of_range_default_rejected(tmp_path: Path) -> None:
    # control_weight default below its declared min must fail validation at load time.
    doc = yaml.safe_load((CONFIG_DIR / f"{_ALGO}.yaml").read_text())
    for entry in doc["params"]:
        if entry["name"] == "control_weight":
            entry["default"] = 0.0  # below min 0.001
    bad = tmp_path / f"{_ALGO}.yaml"
    bad.write_text(yaml.safe_dump(doc), encoding="utf-8")
    with pytest.raises(ParamError):
        ParamSet.from_yaml(bad)


def test_optimal_cost_is_a_metric_contract() -> None:
    r = 1.0
    at_rest = (1.0, 2.0, 0.0, 0.0)
    # Coincident states cost nothing (τ*=0), the identity of the cost geometry.
    cost_self, tau_self = optimal_cost(at_rest, at_rest, r)
    assert cost_self == 0.0
    assert tau_self == 0.0
    # Distinct states cost strictly more than zero, reached at a positive arrival time.
    for x1 in [(4.0, 6.0, 0.0, 0.0), (1.0, 2.0, 1.0, -1.0), (0.0, 0.0, 0.5, 0.5)]:
        cost, tau = optimal_cost(at_rest, x1, r)
        assert cost > 0.0
        assert tau > 0.0


def test_higher_control_weight_costs_more() -> None:
    # J = ∫ 1 + r·uᵀu dt: a larger control weight penalises the same maneuver more.
    x0, x1 = (0.0, 0.0, 0.0, 0.0), (3.0, 2.0, 0.0, 0.0)
    cheap, _ = optimal_cost(x0, x1, 0.5)
    dear, _ = optimal_cost(x0, x1, 5.0)
    assert dear > cheap

File: cpp/tests/test_kinodynamic_rrt_star.cpp
#include <cmath>
#include <string>
#include <vector>

#include <gtest/gtest.h>

#include "navigation/core/params.hpp"
#include "navigation/global_planning/sampling/kinodynamic_rrt_star.hpp"
#include "test_util.hpp"

using namespace navigation;
using core::Point;

namespace {

// Open square grid (resolution 0.5): '.' everywhere.
std::vector<std::string> open_rows(int n) {
  return std::vector<std::string>(static_cast<size_t>(n), std::string(static_cast<size_t>(n), '.'));
}

std::string kino_cfg(int max_iterations, double goal_tolerance) {
  return test::write_temp(
      "kinodynamic_rrt_star.yaml",
      "algorithm: kinodynamic_rrt_star\ncategory: global_planning\nparams:\n"
      "  - {name: max_iterations, type: int, default: " + std::to_string(max_iterations) +
          ", min: 1, max: 200000, description: n}\n"
      "  - {name: goal_bias, type: float, default: 0.2, min: 0.0, max: 1.0, description: b}\n"
      "  - {name: goal_tolerance, type: float, default: " + std::to_string(goal_tolerance) +
          ", min: 0.0, max: 100, description: t}\n"
      "  - {name: neighbor_radius, type: float, default: 2.0, min: 0.01, max: 100, description: r}\n"
      "  - {name: control_weight, type: float, default: 1.0, min: 0.001, max: 1000, description: w}\n"
      "  - {name: max_velocity, type: float, default: 1.5, min: 0.01, max: 100, description: v}\n"
      "  - {name: footprint_radius, type: float, default: 0.15, min: 0.01, max: 20, description: fp}\n"
      "  - {name: seed, type: int, default: 3, min: 0, max: 2147483647, description: s}\n");
}

}  // namespace

// (a) valid, dynamically-feasible, collision-free path on an open grid --------

TEST(KinodynamicRrtStar, FindsFeasiblePathOnOpenGrid) {
  auto g = test::make_grid(open_rows(12));  // 6m x 6m world
  Point start{0.75, 0.75}, goal{5.25, 5.25};
  global_planning::KinodynamicRrtStarPlanner p(core::ParamSet::from_yaml(kino_cfg(4000, 1.5)));
  auto r = p.plan(g, start, goal, nullptr);
  ASSERT_TRUE(r.success);
  // Optimal steering arrives at the goal rest-state; the projected endpoint is within
  // tolerance of the goal position.
  EXPECT_LE(std::hypot(r.path.back().x - goal.x, r.path.back().y - goal.y), 1.5);
  // The propagated double-integrator trajectory must be collision-free throughout.
  for (size_t i = 0; i + 1 < r.path.size(); ++i) {
    EXPECT_TRUE(g.is_motion_valid(r.path[i], r.path[i + 1])) << "path segment must be collision-free";
  }
  EXPECT_GT(r.cost, 0.0);
}

// (b) no-path case: a full wall separates start from goal --------------------

TEST(KinodynamicRrtStar, NoPathAcrossFullWall) {
  auto rows = open_rows(10);
  rows[5] = "##########";  // impassable horizontal wall
  auto g = test::make_grid(rows);
  Point start{0.25, 0.25}, goal{4.75, 4.75};
  global_planning::KinodynamicRrtStarPlanner p(core::ParamSet::from_yaml(kino_cfg(300, 1.0)));
  auto r = p.plan(g, start, goal, nullptr);
  EXPECT_FALSE(r.success);
  EXPECT_TRUE(r.path.empty());
}

// (c) param validation failure: out-of-range default -------------------------

TEST(KinodynamicRrtStar, BadControlWeightThrows) {
  std::string bad = test::write_temp(
      "kinodynamic_rrt_star.yaml",
      "algorithm: kinodynamic_rrt_star\ncategory: global_planning\nparams:\n"
      "  - {name: control_weight, type: float, default: 0.0, min: 0.001, max: 1000, description: under min}\n");
  EXPECT_THROW(core::ParamSet::from_yaml(bad), std::runtime_error);
}
